### PR TITLE
torch.tensor: add tests for list of numpy arrays case

### PR DESCRIPTION
**Pull request by kshitij12345**
References: https://github.com/pytorch/pytorch/issues/13918

Add more test cases for list of numpy array inputs

### CODE REVIEW

File: torch/csrc/utils/tensor_new.cpp [hunk 1]
```diff
@@ -204,10 +204,32 @@ void recursive_store(char* data, IntArrayRef sizes, IntArrayRef strides, int64_t
...
-      TORCH_WARN_ONCE(
-        "Creating a tensor from a list of numpy.ndarrays is extremely slow. "
-        "Please consider converting the list to a single numpy.ndarray with "
-        "numpy.array() before converting to a tensor.");
+      auto np_arr = (PyArrayObject*)items[i];
+      auto np_type = numpy_dtype_to_aten(PyArray_TYPE(np_arr));
+      auto np_arr_is_contiguous = PyArray_IS_C_CONTIGUOUS(np_arr);
+      if (np_arr_is_contiguous) {
+        // take fast-path for contiguous array.
+        TORCH_CHECK(at::can_cast(np_type, scalarType), "Can't cast from ", np_type, " to ", scalarType);
```
Comment (kshitij12345) [inline torch/csrc/utils/tensor_new.cpp:212]: Should be a warning in Python <= 3.9 and an error in Python 3.10 onwards

File: test/test_numpy_interop.py [hunk 2]
```diff
@@ -229,9 +229,9 @@ def test_from_numpy(self, device) -> None:
         self.assertRaises(ValueError, lambda: torch.from_numpy(x))
 
     def test_from_list_of_ndarray_warning(self, device):
-        warning_msg = r"Creating a tensor from a list of numpy.ndarrays is extremely slow"
+        warning_msg = r"Creating a tensor from a list of non-contiguous numpy.ndarrays is extremely slow"
```
Comment (mruberry) [inline test/test_numpy_interop.py:232]: Nice update

File: torch/csrc/utils/tensor_new.cpp [hunk 3]
```diff
@@ -204,10 +204,32 @@ void recursive_store(char* data, IntArrayRef sizes, IntArrayRef strides, int64_t
   for(const auto i : c10::irange(n)) {
 #ifdef USE_NUMPY
     if (is_numpy_available() && PyArray_Check(items[i])) {
-      TORCH_WARN_ONCE(
-        "Creating a tensor from a list of numpy.ndarrays is extremely slow. "
-        "Please consider converting the list to a single numpy.ndarray with "
-        "numpy.array() before converting to a tensor.");
+      auto np_arr = (PyArrayObject*)items[i];
```
Comment (mruberry) [inline torch/csrc/utils/tensor_new.cpp:207]: Is this guaranteed to be a NumPy array? Do we have to deal with cases like a list of lists of lists? Is our testing of this functionality very thorough?

**Review by mruberry — COMMENTED** [review]
Thanks for your patience with this PR, @kshitij12345. I should be more responsive now.

The approach taken here seems reasonable and the speedup is great!

How thorough is our testing for these operations? Would it detect if an edge case is broken?

**Comment by ngimel** [comment]
This PR doesn't do proper size checks.
On master:
```
In [7]: torch.tensor([np.random.random(size=(3, 3)), np.random.random(size=(2, 2))], device=device).shape
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-7-805d1ba0da90> in <module>
----> 1 torch.tensor([np.random.random(size=(3, 3)), np.random.random(size=(2, 2))], device=device).shape

ValueError: expected sequence of length 3 at dim 1 (got 2)
```
On this PR:
```
In [4]: torch.tensor([np.random.random(size=(3, 3)), np.random.random(size=(2, 2))], device=device).shape
Out[4]: torch.Size([2, 3, 3])
```
So test coverage is definitely not enough.

**Comment by kshitij12345** [comment]
@mruberry Have removed the changes and just kept the new tests.

**Review by mruberry — APPROVED** [review]
Cool!

**Pull request closed without merging**